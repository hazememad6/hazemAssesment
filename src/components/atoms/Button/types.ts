import { TouchableOpacityProps } from "react-native";

export interface ButtonProps extends Omit<TouchableOpacityProps, "onPress"> {
  title?: string;
  onPress: () => void;
  variant?: "primary" | "secondary" | "outline" | "ghost";
  size?: "small" | "medium" | "large";
  disabled?: boolean;
  loading?: boolean;
  icon?: string;
  iconOnly?: boolean;
}
