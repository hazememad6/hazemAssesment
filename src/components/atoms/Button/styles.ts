import { StyleSheet } from "react-native";
import { useMetrics } from "@hooks/useMetrics";
import { useThemeStore } from "@store/themeStore";
import { Dimensions } from "@utils";

interface UseButtonStylesProps {
  variant: "primary" | "secondary" | "outline" | "ghost";
  size: "small" | "medium" | "large";
  disabled: boolean;
  iconOnly?: boolean;
}

export const useButtonStyles = ({ variant, size, disabled, iconOnly }: UseButtonStylesProps) => {
  const { moderateScale, horizontalScale, verticalScale } = useMetrics();
  const { theme } = useThemeStore();

  const getButtonPadding = () => {
    if (iconOnly) {
      switch (size) {
        case "small":
          return moderateScale(Dimensions.primarySpace);
        case "large":
          return moderateScale(Dimensions.sectionSpace);
        default:
          return moderateScale(Dimensions.secondarySpace);
      }
    }

    switch (size) {
      case "small":
        return {
          paddingHorizontal: horizontalScale(Dimensions.secondarySpace),
          paddingVertical: verticalScale(Dimensions.primarySpace),
        };
      case "large":
        return {
          paddingHorizontal: horizontalScale(Dimensions.extraSpace),
          paddingVertical: verticalScale(Dimensions.sectionSpace),
        };
      default:
        return {
          paddingHorizontal: horizontalScale(Dimensions.sectionSpace),
          paddingVertical: verticalScale(Dimensions.secondarySpace),
        };
    }
  };

  const getBackgroundColor = () => {
    if (disabled) return theme.colors.border;
    switch (variant) {
      case "primary":
        return theme.colors.primary;
      case "secondary":
        return theme.colors.secondary;
      case "outline":
      case "ghost":
        return "transparent";
      default:
        return theme.colors.primary;
    }
  };

  const getBorderColor = () => {
    if (variant === "outline") return theme.colors.primary;
    return "transparent";
  };

  const getTextColor = () => {
    if (disabled) return theme.colors.border;
    if (variant === "outline" || variant === "ghost") return theme.colors.primary;
    return "#ffffff";
  };

  const padding = getButtonPadding();
  const flexDirection = iconOnly ? "column" : "row";

  const buttonStyle = {
    backgroundColor: getBackgroundColor(),
    borderRadius: moderateScale(iconOnly ? 20 : Dimensions.borderRadius),
    borderWidth: variant === "outline" ? moderateScale(1) : 0,
    borderColor: getBorderColor(),
    alignItems: "center" as const,
    justifyContent: "center" as const,
    flexDirection: flexDirection as "row" | "column",
    opacity: disabled ? 0.6 : 1,
  };

  // Handle padding for icon-only vs regular buttons
  const finalButtonStyle = iconOnly
    ? {
        ...buttonStyle,
        padding: padding as number,
        width: moderateScale(44),
        height: moderateScale(44),
      }
    : {
        ...buttonStyle,
        ...(padding as object),
      };

  return StyleSheet.create({
    buttonStyle: finalButtonStyle,
    textStyle: {
      fontSize: moderateScale(
        size === "small" ? Dimensions.smallFont : size === "large" ? Dimensions.largeFont : Dimensions.regularFont
      ),
      fontWeight: "600" as const,
      color: getTextColor(),
    },
  });
};
