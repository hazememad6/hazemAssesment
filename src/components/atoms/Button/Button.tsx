import React from "react";
import { ActivityIndicator, Text, TouchableOpacity } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import { useMetrics } from "@hooks/useMetrics";
import { useThemeStore } from "@store/themeStore";
import { Dimensions } from "@utils";
import { useButtonStyles } from "./styles";
import { ButtonProps } from "./types";

export const Button: React.FC<ButtonProps> = ({
  title,
  onPress,
  variant = "primary",
  size = "medium",
  disabled = false,
  loading = false,
  icon,
  iconOnly = false,
  ...rest
}) => {
  const { buttonStyle, textStyle } = useButtonStyles({
    variant,
    size,
    disabled,
    iconOnly,
  });
  const { theme } = useThemeStore();
  const { moderateScale } = useMetrics();

  const getIconSize = () => {
    switch (size) {
      case "small":
        return moderateScale(Dimensions.smallIcon);
      case "large":
        return moderateScale(Dimensions.largeIcon);
      default:
        return moderateScale(Dimensions.regularIcon);
    }
  };

  const getIconColor = () => {
    if (disabled) return theme.colors.border;
    if (variant === "outline" || variant === "ghost") return theme.colors.primary;
    return "#ffffff";
  };

  const renderContent = () => {
    if (loading) {
      return (
        <ActivityIndicator
          size="small"
          color={variant === "outline" || variant === "ghost" ? theme.colors.primary : "#ffffff"}
        />
      );
    }

    if (iconOnly && icon) {
      return <Ionicons name={icon as any} size={getIconSize()} color={getIconColor()} testID={`icon-${icon}`} />;
    }

    if (icon && title) {
      return (
        <>
          <Ionicons
            name={icon as any}
            size={getIconSize()}
            color={getIconColor()}
            style={{ marginRight: moderateScale(Dimensions.primarySpace) }}
            testID={`icon-${icon}`}
          />
          <Text style={textStyle}>{title}</Text>
        </>
      );
    }

    return <Text style={textStyle}>{title}</Text>;
  };

  return (
    <TouchableOpacity
      style={buttonStyle}
      onPress={onPress}
      disabled={disabled || loading}
      activeOpacity={0.7}
      accessibilityRole="button"
      {...rest}
    >
      {renderContent()}
    </TouchableOpacity>
  );
};
