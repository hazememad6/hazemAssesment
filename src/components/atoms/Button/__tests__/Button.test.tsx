import React from "react";
import { fireEvent, render } from "@testing-library/react-native";
import { Button } from "../Button";

describe("Button Component", () => {
  const defaultProps = {
    title: "Test Button",
    onPress: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("Snapshots", () => {
    it("renders primary button correctly", () => {
      const tree = render(<Button {...defaultProps} variant="primary" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders secondary button correctly", () => {
      const tree = render(<Button {...defaultProps} variant="secondary" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders ghost button correctly", () => {
      const tree = render(<Button {...defaultProps} variant="ghost" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders disabled button correctly", () => {
      const tree = render(<Button {...defaultProps} disabled />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders icon-only button correctly", () => {
      const tree = render(<Button onPress={defaultProps.onPress} icon="add" iconOnly />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders button with icon and title correctly", () => {
      const tree = render(<Button {...defaultProps} icon="heart" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Behavior", () => {
    it("calls onPress when button is pressed", () => {
      const { getByText } = render(<Button {...defaultProps} />);
      const button = getByText("Test Button");

      fireEvent.press(button);

      expect(defaultProps.onPress).toHaveBeenCalledTimes(1);
    });

    it("does not call onPress when button is disabled", () => {
      const { getByText } = render(<Button {...defaultProps} disabled />);
      const button = getByText("Test Button");

      fireEvent.press(button);

      expect(defaultProps.onPress).not.toHaveBeenCalled();
    });

    it("renders correct testID", () => {
      const { getByTestId } = render(<Button {...defaultProps} testID="custom-button" />);

      expect(getByTestId("custom-button")).toBeTruthy();
    });

    it("renders icon when icon prop is provided", () => {
      const { getByTestId } = render(<Button {...defaultProps} icon="add" />);

      expect(getByTestId("icon-add")).toBeTruthy();
    });

    it("does not render title when iconOnly is true", () => {
      const { queryByText } = render(<Button onPress={defaultProps.onPress} icon="add" iconOnly />);

      expect(queryByText("Test Button")).toBeNull();
    });

    it("renders both icon and title when iconOnly is false", () => {
      const { getByText, getByTestId } = render(<Button {...defaultProps} icon="add" />);

      expect(getByText("Test Button")).toBeTruthy();
      expect(getByTestId("icon-add")).toBeTruthy();
    });
  });

  describe("Accessibility", () => {
    it("has correct accessibility properties", () => {
      const { getByRole } = render(<Button {...defaultProps} />);
      const button = getByRole("button");

      expect(button).toBeTruthy();
      expect(button.props.accessible).toBe(true);
    });

    it("has correct accessibility label", () => {
      const { getByLabelText } = render(<Button {...defaultProps} accessibilityLabel="Custom Label" />);

      expect(getByLabelText("Custom Label")).toBeTruthy();
    });
  });

  describe("Edge Cases", () => {
    it("handles undefined title gracefully", () => {
      const result = render(<Button onPress={defaultProps.onPress} />);

      expect(result).toBeTruthy();
    });

    it("handles multiple rapid presses", () => {
      const { getByText } = render(<Button {...defaultProps} />);
      const button = getByText("Test Button");

      fireEvent.press(button);
      fireEvent.press(button);
      fireEvent.press(button);

      expect(defaultProps.onPress).toHaveBeenCalledTimes(3);
    });
  });
});
