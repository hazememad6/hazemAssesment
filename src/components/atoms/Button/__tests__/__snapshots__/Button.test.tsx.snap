// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Snapshots renders button with icon and title correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#007AFF",
      "borderColor": "transparent",
      "borderRadius": undefined,
      "borderWidth": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    color="#ffffff"
    size={24}
    style={
      {
        "marginRight": 16,
      }
    }
    testID="icon-heart"
  >
    Icon-heart
  </Text>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontSize": undefined,
        "fontWeight": "600",
      }
    }
  >
    Test Button
  </Text>
</View>
`;

exports[`Button Component Snapshots renders disabled button correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#C6C6C8",
      "borderColor": "transparent",
      "borderRadius": undefined,
      "borderWidth": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 0.6,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#C6C6C8",
        "fontSize": undefined,
        "fontWeight": "600",
      }
    }
  >
    Test Button
  </Text>
</View>
`;

exports[`Button Component Snapshots renders ghost button correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "transparent",
      "borderRadius": undefined,
      "borderWidth": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#007AFF",
        "fontSize": undefined,
        "fontWeight": "600",
      }
    }
  >
    Test Button
  </Text>
</View>
`;

exports[`Button Component Snapshots renders icon-only button correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#007AFF",
      "borderColor": "transparent",
      "borderRadius": 20,
      "borderWidth": 0,
      "flexDirection": "column",
      "height": 44,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": 44,
    }
  }
>
  <Text
    color="#ffffff"
    size={24}
    testID="icon-add"
  >
    Icon-add
  </Text>
</View>
`;

exports[`Button Component Snapshots renders primary button correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#007AFF",
      "borderColor": "transparent",
      "borderRadius": undefined,
      "borderWidth": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontSize": undefined,
        "fontWeight": "600",
      }
    }
  >
    Test Button
  </Text>
</View>
`;

exports[`Button Component Snapshots renders secondary button correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#5856D6",
      "borderColor": "transparent",
      "borderRadius": undefined,
      "borderWidth": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontSize": undefined,
        "fontWeight": "600",
      }
    }
  >
    Test Button
  </Text>
</View>
`;
